Settings.c

```diff
@@ -148,51 +196,107 @@ static void Settings_defaultMeters(Settings* this, unsigned int initialCpuCount)
+      return column->name;
+   }
+   return Process_fields[id].name;
+}
+
+static int toFieldIndex(Hashtable* columns, const char* str) {
+   if (isdigit(str[0])) {
+      // This "+1" is for compatibility with the older enum format.
+      int id = atoi(str) + 1;
+      if (toFieldName(columns, id)) {
+         return id;
+      }
+   } else {
+      // Dynamically-defined columns are always stored by-name.
+      char dynamic[32] = {0};
+      if (sscanf(str, "Dynamic(%30s)", dynamic)) {
+         char* end;
+         if ((end = strrchr(dynamic, ')')) != NULL) {
+            bool success;
+            unsigned int key;
+            *end = '\0';
+            success = DynamicColumn_search(columns, dynamic, &key) != NULL;
+            *end = ')';
+            if (success)
+               return key;
+         }
+      }
+      // Fallback to iterative scan of table of fields by-name.
+      for (int p = 1; p < LAST_PROCESSFIELD; p++) {
+         const char* pName = toFieldName(columns, p);
+         if (pName && strcmp(pName, str) == 0)
+            return p;
+      }
+   }
+   return -1;
+}
+
+static void ScreenSettings_readFields(ScreenSettings* ss, Hashtable* columns, const char* line) {
    char* trim = String_trim(line);
    char** ids = String_split(trim, ' ', NULL);
    free(trim);
 
-   settings->flags = 0;
-
    unsigned int i, j;
    for (j = 0, i = 0; ids[i]; i++) {
       if (j >= UINT_MAX / sizeof(ProcessField))
          continue;
       if (j >= LAST_PROCESSFIELD) {
-         settings->fields = xRealloc(settings->fields, j * sizeof(ProcessField));
-         memset(&settings->fields[j], 0, sizeof(ProcessField));
-      }
-
-      // Dynamically-defined columns are always stored by-name.
-      char dynamic[32] = {0};
-      if (sscanf(ids[i], "Dynamic(%30s)", dynamic)) {
-         char* end;
-         if ((end = strrchr(dynamic, ')')) == NULL)
-            continue;
-         *end = '\0';
-         unsigned int key;
-         if (!DynamicColumn_search(settings->dynamicColumns, dynamic, &key))
-            continue;
-         settings->fields[j++] = key;
-         continue;
-      }
-      // This "+1" is for compatibility with the older enum format.
-      int id = atoi(ids[i]) + 1;
-      if (id > 0 && id < LAST_PROCESSFIELD && Process_fields[id].name) {
-         settings->flags |= Process_fields[id].flags;
-         settings->fields[j++] = id;
+         ss->fields = xRealloc(ss->fields, (j + 1) * sizeof(ProcessField));
+         memset(&ss->fields[j], 0, sizeof(ProcessField));
       }
+      int id = toFieldIndex(columns, ids[i]);
+      if (id >= 0)
+         ss->fields[j] = id;
+      if (id > 0 && id < LAST_PROCESSFIELD)
+         ss->flags |= Process_fields[id].flags;
+      j++;
    }
-   settings->fields[j] = NULL_PROCESSFIELD;
    String_freeArray(ids);
 }
 
+ScreenSettings* Settings_newScreen(Settings* this, const char* name, const char* line) {
+   ScreenSettings* ss = xCalloc(sizeof(ScreenSettings), 1);
```

Comment:
I think this pointer didn't free properly. I will try to debug this later, 
ASan output:
```
==195406==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 40 byte(s) in 1 object(s) allocated from:
    #0 0x7f4adc282dc6 in calloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10ddc6)
    #1 0x559095bdcb64 in xCalloc /home/smalinux/htop/XUtils.c:55
    #2 0x559095bd3546 in Settings_newScreen /home/smalinux/htop/Settings.c:266
    #3 0x559095bd38dd in Settings_defaultScreens /home/smalinux/htop/Settings.c:289
    #4 0x559095bd3b2b in Settings_defaultScreens /home/smalinux/htop/Settings.c:285
    #5 0x559095bd3b2b in Settings_read /home/smalinux/htop/Settings.c:328
    #6 0x559095bd7fbc in Settings_new /home/smalinux/htop/Settings.c:732
    #7 0x559095b9f988 in CommandLine_run /home/smalinux/htop/CommandLine.c:303
    #8 0x7f4adbdc30b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

SUMMARY: AddressSanitizer: 40 byte(s) leaked in 1 allocation(s).
```

Reply:
`xCalloc` like `calloc` takes the number of elements first and the size of each element as second. Mostly a side-note, but having them swapped may lead to unnecessary trouble down the line.

That's most likely not the bug you're chasing, but you're advised to fix it while at it. ;-)

Reply:
@smalinux thanks, I have a fix now - will push shortly.